File: test/Projection/TestProjection.f90
program TestProjection
  use Projection
  implicit none


  integer :: iorb1,iorb2,is,ik,iomega,is1,is2
  complex*16 :: tempmat1(4,3), err, &
    glatt(4,4,2,5,0:9), gproj(4,3,2,5), glocref(3,3,2,5,0:9), &
    glocstc(3,3,2), glocdyn(3,3,2,0:9), glatstc(3,3,2,5), glatdyn(3,3,2,5,0:9), &
    wlatt(4,4,2,2,5,0:9), wproj(4,3,2,5), wlocref(3,3,2,2,5,0:9), &
    wlocstc(3,3,2,2), wlocdyn(3,3,2,2,0:9), wlatstc(3,3,2,2,5), wlatdyn(3,3,2,2,5,0:9)  


  glatt=0.0d0
  glocref=0.0d0
  gproj=0.0d0

  do iomega=0, 9
    do ik=1, 5
      do is=1, 2
        do iorb1=1, 4
          do iorb2=1, 4
            glatt(iorb1,iorb2,is,ik,iomega)=10*ik+iorb1+0.1d0*iomega+iorb2*2+dcmplx(0.0d0,(is)*0.1d0)
          enddo
        enddo
      enddo
    enddo
  enddo


  do ik=1, 5      
    do is=1, 2
      do iorb1=1, 4
        do iorb2=1, 3
          gproj(iorb1,iorb2,is,ik)=0.1d0*is+ik*0.5d0+(iorb1-iorb2)*2.0d0
        enddo
      enddo
    enddo
  enddo


  do iomega=0, 9
    do ik=1, 5      
      do is=1, 2
        tempmat1=matmul(glatt(:,:,is,ik,iomega), gproj(:,:,is,ik))
        glocref(:,:,is,ik,iomega)=glocref(:,:,is,ik,iomega)+matmul(transpose(dconjg(gproj(:,:,is,ik))), tempmat1)
      enddo
    enddo
  enddo

  call FLocStc(4,2,glatt(:,:,:,1,0), 3,gproj(:,:,:,1),glocstc)
  call FLatStc(4,2,5,glatt(:,:,:,:,0), 3,gproj,glatstc)
  call FLocDyn(4,2,10,glatt(:,:,:,1,:), 3,gproj(:,:,:,1),glocdyn)
  call FLatDyn(4,2,5,10,glatt,3,gproj,glatdyn)


  do is=1, 2
    do iorb1=1, 3
      do iorb2=1, 3
        err=glocref(iorb1,iorb2,is,1,0)-glocstc(iorb1,iorb2,is)
        if (cdabs(err) .gt. 1.0d-6) then
          print '(3i5, 5(2x, f20.12))', iorb1, iorb2, is, cdabs(err), glocref(iorb1,iorb2,is,1,0), glocstc(iorb1,iorb2,is)
        endif
      enddo
    enddo
  enddo

  do ik=1, 5
    do is=1, 2
      do iorb1=1, 3
        do iorb2=1, 3      
          err=glocref(iorb1,iorb2,is,ik,0)-glatstc(iorb1,iorb2,is,ik)
          if (cdabs(err) .gt. 1.0d-6) then
            print '(4i5, 5(2x, f12.6))', iorb1, iorb2, is, ik,cdabs(err), glocref(iorb1,iorb2,is,ik,0), glatstc(iorb1,iorb2,is,ik)
          endif
        enddo
      enddo
    enddo
  enddo

  do iomega=0, 9
    do is=1, 2
      do iorb1=1, 3
        do iorb2=1, 3      
          err=glocref(iorb1,iorb2,is,1,iomega)-glocdyn(iorb1,iorb2,is,iomega)
          if (cdabs(err) .gt. 1.0d-6) then
            print '(4i5, 5(2x, f12.6))', iorb1, iorb2, is, iomega, cdabs(err), glocref(iorb1,iorb2,is,1,iomega), glocdyn(iorb1,iorb2,is,iomega)
          endif
        enddo
      enddo
    enddo
  enddo


  do iomega=0, 9
    do ik=1, 5
      do is=1, 2
        do iorb1=1, 3
          do iorb2=1, 3      
            err=glocref(iorb1,iorb2,is,ik,iomega)-glatdyn(iorb1,iorb2,is,ik,iomega)
            if (cdabs(err) .gt. 1.0d-6) then
              print '(5i5, 5(2x, f12.6))', iorb1, iorb2, is, ik,iomega, cdabs(err), glocref(iorb1,iorb2,is,ik,iomega), glatdyn(iorb1,iorb2,is,ik,iomega)
            endif
          enddo
        enddo
      enddo
    enddo
  enddo



  wlatt=0.0d0
  wlocref=0.0d0
  wproj=0.0d0

  do iomega=0, 9
    do ik=1, 5
      do is1=1, 2
        do is2=1, 2        
          do iorb1=1, 4
            do iorb2=1, 4
              wlatt(iorb1,iorb2,is1,is2,ik,iomega)=10*ik+iorb1+0.1d0*iomega+iorb2*2+dcmplx(0.0d0,(is1-is2)*0.1d0)
            enddo
          enddo
        enddo
      enddo
    enddo
  enddo


  do ik=1, 5      
    do is1=1, 2
      do iorb1=1, 4
        do iorb2=1, 3
          wproj(iorb1,iorb2,is1,ik)=0.1d0*is1+ik*0.5d0+(iorb1-iorb2)*2.0d0
        enddo
      enddo
    enddo
  enddo


  do iomega=0, 9
    do ik=1, 5      
      do is1=1, 2
        do is2=1, 2        
          tempmat1=matmul(wlatt(:,:,is1,is2,ik,iomega), wproj(:,:,is2,ik))
          wlocref(:,:,is1,is2,ik,iomega)=wlocref(:,:,is1,is2,ik,iomega)+matmul(transpose(dconjg(wproj(:,:,is1,ik))), tempmat1)
        enddo
      enddo
    enddo
  enddo

  call BLocStc(4,2,wlatt(:,:,:,:,1,0), 3,wproj(:,:,:,1),wlocstc)
  call BLatStc(4,2,5,wlatt(:,:,:,:,:,0), 3,wproj,wlatstc)
  call BLocDyn(4,2,10,wlatt(:,:,:,:,1,:), 3,wproj(:,:,:,1),wlocdyn)
  call BLatDyn(4,2,5,10,wlatt,3,wproj,wlatdyn)

  print *, 'BLocStc'
  do is1=1, 2
    do is2=1, 2      
      do iorb1=1, 3
        do iorb2=1, 3
          err=wlocref(iorb1,iorb2,is1,is2,1,0)-wlocstc(iorb1,iorb2,is1,is2)
          if (cdabs(err) .gt. 1.0d-6) then
            print '(4i5, 5(2x, f20.12))', iorb1, iorb2, is1, is2,cdabs(err), wlocref(iorb1,iorb2,is1,is2,1,0), wlocstc(iorb1,iorb2,is1,is2)
          endif
        enddo
      enddo
    enddo
  enddo

  print *, 'BLatStc'  
  do ik=1, 5
    do is1=1, 2
      do is2=1, 2      
        do iorb1=1, 3
          do iorb2=1, 3      
            err=wlocref(iorb1,iorb2,is1,is2,ik,0)-wlatstc(iorb1,iorb2,is1,is2,ik)
            if (cdabs(err) .gt. 1.0d-6) then
              print '(5i5, 5(2x, f12.6))', iorb1, iorb2, is1, is2,ik,cdabs(err), wlocref(iorb1,iorb2,is1,is2,ik,0), wlatstc(iorb1,iorb2,is1,is2,ik)
            endif
          enddo
        enddo
      enddo
    enddo
  enddo

  print *, 'BLocDyn'  
  do iomega=0, 9
    do is1=1, 2
      do is2=1, 2      
        do iorb1=1, 3
          do iorb2=1, 3      
            err=wlocref(iorb1,iorb2,is1,is2,1,iomega)-wlocdyn(iorb1,iorb2,is1,is2,iomega)
            if (cdabs(err) .gt. 1.0d-6) then
              print '(5i5, 5(2x, f12.6))', iorb1, iorb2, is1, is2,iomega, cdabs(err), wlocref(iorb1,iorb2,is1,is2,1,iomega), wlocdyn(iorb1,iorb2,is1,is2,iomega)
            endif
          enddo
        enddo
      enddo
    enddo
  enddo

  print *, 'BLatDyn'  
  do iomega=0, 9
    do ik=1, 5
      do is1=1, 2
        do is2=1, 2        
          do iorb1=1, 3
            do iorb2=1, 3      
              err=wlocref(iorb1,iorb2,is1,is2,ik,iomega)-wlatdyn(iorb1,iorb2,is1,is2,ik,iomega)
              if (cdabs(err) .gt. 1.0d-6) then
                print '(6i5, 5(2x, f12.6))', iorb1, iorb2, is1, is2,ik,iomega, cdabs(err), wlocref(iorb1,iorb2,is1,is2,ik,iomega), wlatdyn(iorb1,iorb2,is1,is2,ik,iomega)
              endif
            enddo
          enddo
        enddo
      enddo
    enddo
  enddo
end program TestProjection

File: class/test/TestProjection.py
import numpy as np
import DiagE

norb = 4
norbc = 3
ns = 2
nk = 5
nomega = 10

tempmat1 = np.zeros([norb, norbc],dtype=complex,order='F')
glatt = np.zeros([norb,norb,ns,nk,nomega],dtype=complex,order='F')
gproj = np.zeros([norb,norbc,ns,nk],dtype=complex,order='F')
glocref = np.zeros([norbc,norbc,ns,nk,nomega],dtype=complex,order='F')
glocstc = np.zeros([norbc,norbc,ns],dtype=complex,order='F')
glocdyn = np.zeros([norbc,norbc,ns,nomega],dtype=complex,order='F')
glatstc = np.zeros([norbc,norbc,ns,nk],dtype=complex,order='F')
glatdyn = np.zeros([norbc,norbc,ns,nk,nomega],dtype=complex,order='F')
wlatt = np.zeros([norb,norb,ns,ns,nk,nomega],dtype=complex,order='F')
wproj = np.zeros([norb,norbc,ns,nk],dtype=complex,order='F')
wlocref = np.zeros([norbc,norbc,ns,ns,nk,nomega],dtype=complex,order='F')
wlocstc = np.zeros([norbc,norbc,ns,ns],dtype=complex,order='F')
wlocdyn = np.zeros([norbc,norbc,ns,ns,nomega],dtype=complex,order='F')
wlatstc = np.zeros([norbc,norbc,ns,ns,nk],dtype=complex,order='F')
wlatdyn = np.zeros([norbc,norbc,ns,ns,nk,nomega],dtype=complex,order='F')

for iomega in range(nomega):
    for ik in range(nk):
        for js in range(ns):
            for iorb in range(norb):
                for jorb in range(norb):
                    glatt[iorb,jorb,js,ik,iomega] = 10*(ik+1)+(iorb+1)+0.1*(iomega+1)+(jorb+1)*2+1j*(js+1)*0.1


for ik in range(nk):
    for js in range(ns):
        for iorb in range(norb):
            for jorb in range(norbc):
                gproj[iorb,jorb,js,ik] = 0.1*(js+1) +(ik+1)*0.5 +((iorb+1)-(jorb+1))*2.0


for iomega in range(nomega):
    for ik in range(nk):
        for js in range(ns):
            tempmat1 = np.matmul(glatt[:,:,js,ik,iomega],gproj[:,:,js,ik])
            glocref[:,:,js,ik,iomega] += np.matmul(gproj[:,:,js,ik].T.conj(),tempmat1)

glocstc = DiagE.projection.flocstc(glatt[:,:,:,0,0],gproj[:,:,:,0])
glatstc = DiagE.projection.flatstc(glatt[:,:,:,:,0],gproj)
glocdyn = DiagE.projection.flocdyn(glatt[:,:,:,0,:],gproj[:,:,:,0])
glatdyn = DiagE.projection.flatdyn(glatt,gproj)


print("Projectoin FLocStc")

for js in range(ns):
    for iorb in range(norbc):
        for jorb in range(norbc):
            err = glocstc[iorb,jorb,js] - glocref[iorb,jorb,js,0,0]
            if abs(err) >= 1.0e-6:
                print(iorb,jorb,js,abs(err),glocstc[iorb,jorb,js], glocref[iorb,jorb,js,0,0])


print("Projection FLatStc")

for ik in range(nk):
    for js in range(ns):
        for iorb in range(norbc):
            for jorb in range(norbc):
                err = glatstc[iorb,jorb,js,ik]-glocref[iorb,jorb,js,ik,0]
                if abs(err) >=1.0e-6:
                    print(iorb,jorb,js,ik,abs(err),glatstc[iorb,jorb,js,ik],glocref[iorb,jorb,js,ik,0])

print("Projection FLocDyn")

for iomega in range(nomega):
    for js in range(ns):
        for iorb in range(norbc):
            for jorb in range(norbc):
                err = glocdyn[iorb,jorb,js,iomega]-glocref[iorb,jorb,js,0,iomega]
                if abs(err) >=1.0e-6:
                    print(iorb,jorb,js,iomega,abs(err),glocdyn[iorb,jorb,js,iomega],glocref[iorb,jorb,js,0,iomega])

print("Projection FLatDyn")

for iomega in range(nomega):
    for ik in range(nk):
        for js in range(ns):
            for iorb in range(norbc):
                for jorb in range(norbc):
                    err = glatdyn[iorb,jorb,js,ik,iomega]-glocref[iorb,jorb,js,ik,iomega]
                    if abs(err)>=1.0e-6:
                        print(iorb,jorb,js,ik,iomega,abs(err),glatdyn[iorb,jorb,js,ik,iomega],glocref[iorb,jorb,js,ik,iomega])


for iomega in range(nomega):
    for ik in range(nk):
        for js in range(ns):
            for ks in range(ns):
                for iorb in range(norb):
                    for jorb in range(norb):
                        wlatt[iorb,jorb,js,ks,ik,iomega] = 10*(ik+1)+(iorb+1)*0.1*(iomega+1)+jorb*2+1j*((js+1)-(ks+1))*0.1

for ik in range(nk):
    for js in range(ns):
        for iorb in range(norb):
            for jorb in range(norbc):
                wproj[iorb,jorb,js,ik] = 0.1*(js+1)+(ik+1)*0.5+((iorb+1)-(jorb+1))*2.0

for iomega in range(nomega):
    for ik in range(nk):
        for js in range(ns):
            for ks in range(ns):
                tempmat1 = np.matmul(wlatt[:,:,js,ks,ik,iomega],wproj[:,:,ks,ik])
                wlocref[:,:,js,ks,ik,iomega] += np.matmul(wproj[:,:,js,ik].T.conj(),tempmat1)


wlocstc = DiagE.projection.blocstc(wlatt[:,:,:,:,0,0],wproj[:,:,:,0])
wlatstc = DiagE.projection.blatstc(wlatt[:,:,:,:,:,0],wproj)
wlocdyn = DiagE.projection.blocdyn(wlatt[:,:,:,:,0,:],wproj[:,:,:,0])
wlatdyn = DiagE.projection.blatdyn(wlatt,wproj)

print("Projection BLocStc")

for js in range(ns):
    for ks in range(ns):
        for iorb in range(norbc):
            for jorb in range(norbc):
                err = wlocstc[iorb,jorb,js,ks]-wlocref[iorb,jorb,js,ks,0,0]
                if abs(err) >= 1.0e-6:
                    print(iorb,jorb,js,ks,abs(err),wlocstc[iorb,jorb,js,ks],wlocref[iorb,jorb,js,ks,0,0])

print("Projection BLatStc")

for ik in range(nk):
    for js in range(ns):
        for ks in range(ns):
            for iorb in range(norbc):
                for jorb in range(norbc):
                    err=wlatstc[iorb,jorb,js,ks,ik]-wlocref[iorb,jorb,js,ks,ik,0]
                    if abs(err) >=1.0e-6:
                        print(iorb,jorb,js,ks,ik,abs(err),wlatstc[iorb,jorb,js,ks,ik],wlocref[iorb,jorb,js,ks,ik,0])

print("Projection BLocDyn")

for iomega in range(nomega):
    for js in range(ns):
        for ks in range(ns):
            for iorb in range(norbc):
                for jorb in range(norbc):
                    err = wlocdyn[iorb,jorb,js,ks,iomega]-wlocref[iorb,jorb,js,ks,0,iomega]
                    if abs(err) >= 1.0e-6:
                        print(iorb,jorb,js,ks,iomega,abs(err),wlocdyn[iorb,jorb,js,ks,iomega],wlocref[iorb,jorb,js,ks,0,iomega])

print("Projection BLatDyn")

for iomega in range(nomega):
    for ik in range(nk):
        for js in range(ns):
            for ks in range(ns):
                for iorb in range(norbc):
                    for jorb in range(norbc):
                        err = wlatdyn[iorb,jorb,js,ks,ik,iomega]-wlocref[iorb,jorb,js,ks,ik,iomega]
                        if abs(err) >= 1.0e-6:
                            print(iorb,jorb,js,ks,ik,iomega,abs(err),wlatdyn[iorb,jorb,js,ks,ik,iomega]-wlocref[iorb,jorb,js,ks,ik,iomega])
